### DISCUSSION TITLE
How could I communicate through UART on platform STM32L433? (Ubuntu)

### DISCUSSION THREAD

**Discussion by kp008114** [zephyrproject-rtos/zephyr#45296] [Q&A]
Hello everyone,

  I finshed installing the SDK, west and DTC, then I accomplished the first sample Hello world. I used "screen" to show the words on terminal. It's good. However, I would like to communicate to MCU through UART that means not only MCU can send me some words but also I can send some words to MCU and it can respond something to me based on those words. Unfortunately, I have no idea how to achieve this.

  I used to developing the program in Wondows STMCubeMX, but I have to switch to Linux becasue of some reasons. May some one give me ideas? And if "screen" is a good tool to interact with MCU, may I sending words to MCU through "screen"?

platform: STM32L433
OS: Ubuntu

**Comment by erwango** [comment]
You can test `samples/subsys/shell/shell_module`.
On my side, i'm using minicom instead of screen, but it should be equivalent.

**Reply by kp008114** [reply]
so I can use `west build` command to build `samples/subsys/shell/shell_module`. Am I right?
Or do you mean to use the functions in the `samples/subsys/shell/shell_module/src/main` within my main file?

**Reply by erwango** [reply]
> so I can use west build command to build samples/subsys/shell/shell_module. Am I right? 

Right

**Reply by kp008114** [reply]
But I have to use _**UART**_ for commnication. 
Now PA9 and PA10 are connected to a UART transceiver and the UART transceiver connects to PC on ttyUSB0.
How can I communicate to mcu through it?

**Reply by erwango** [reply]
By default, `samples/subsys/shell/shell_module` will offer a communication link (both RX/TX) over UART.
Have you had a try? If yes, can you explain what is missing regarding the use case you expect?

**Reply by kp008114** [reply]
Finshed build and flash.
Command to connect uart: `sudo screen /dev/ttyUSB0 115200,cs8`
The terminal shows nothing.
Here are some questions:
1. Where can I know which pins the current uart is using? I checked `nucleo_l433rc_p.dts` 
The content as follows. Does that mean I've to hang up the PA9 and PA10 to uart transceiver accordingly?
```&usart1 {
        pinctrl-0 = <&usart1_tx_pa9 &usart1_rx_pa10>;
        pinctrl-names = "default";
        current-speed = <115200>;
```
2. How does the main function know which uart is using? I cannot find any word like &uart1, &uart2, etc. in main.c file.
Is the following line figuring out it?
`dev = DEVICE_DT_GET(DT_CHOSEN(zephyr_shell_uart));`
3. The terminal, where I input `screen command`, didn't show anything. Neither I keyed in some words and pressed Enter.

**Reply by erwango** [reply]
1. Unless you changed it, default  is lpuart1 actually:
```
        zephyr,console = &lpuart1;
        zephyr,shell-uart = &lpuart1;
```

2. Binding is done thanks to `zephyr,shell-uart = &lpuart1;`

3. STLink tty is ACM0, so you should use `screen /dev/ttyACM0 115200,cs8`. This being said, I'm using minicom : `minicom -D /dev/ttyACM0`

**Reply by kp008114** [reply]
I found you're the same person replying me in discord #stm32. Thank you so much.
May I change the default console to uart? Or how can the lpuart(ttyACM0) and uart(ttyUSB0) coexist? Because I have a uart transceiver that is seperated from the usb cable for lpuart.

**Reply by erwango** [reply]
If you have a nucleo_l433rc_p board, I'd suggest making it work the default way to start from a known working basis.
Then, modify as you wish.
